### PR TITLE
Add coalesce fun

### PR DESCRIPTION
**Pull request by ouyuanning**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2357

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

**Review by aressu1985 — APPROVED** [review]

File: pkg/sql/plan/function/operator/coalesce.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+			}
+
+			for j := 0; j < vecLen; j++ {
+				if rs.Nsp.Contains(uint64(j)) && !input.Nsp.Contains(uint64(j)) {
+					rsCols[j] = cols[j]
+					rs.Nsp.Np.Remove(uint64(j))
+				}
+			}
+
+			if rs.Nsp.Np.IsEmpty() {
```
Comment (daviszhen) [inline pkg/sql/plan/function/operator/coalesce.go:164]: An optimization tip,  when the count of not null elements in RS is equal to the vecLen, it denotes that all not null elements have been found and just skips the rest of input vector.

**Review by daviszhen — APPROVED** [review]

File: pkg/sql/plan/function/operator/coalesce.go (hunk 1 above)
Comment (ouyuanning) [reply pkg/sql/plan/function/operator/coalesce.go]: get it

File: pkg/sql/plan/function/operator/coalesce.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+	for i := 0; i < len(vs); i++ {
+		input := vs[i]
+		cols := vector.MustTCols[T](input)
+		if input.IsScalar() {
+			if input.IsScalarNull() {
+				continue
+			}
+
+			if returnIsScalar {
```
Comment (m-schen) [inline pkg/sql/plan/function/operator/coalesce.go:134]: move it out of the loop and before you call the proc.AllocVector ?

Comment (ouyuanning) [reply pkg/sql/plan/function/operator/coalesce.go]: good point

**Review by m-schen — APPROVED** [review]

**Pull request merged**